examples/labs/inverse_kinematics.py

```diff
@@ -0,0 +1,150 @@
+
+
+# *********************************************
+# ** INVERSE KINEMATICS WITH METHOD XXXXXXXXX
+# *********************************************
+
+
+# If jfk is body jacobian, delta_theta is computed by
+#          pose * exp(jfk * delta_theta) = targeted_pose,
+# which has a closed-form solution:
+# .  delta_theta = jfk.pinverse() * log(pose^-1 * targeted_pose)
+# Otherwise, if jfk is spatial jacobian, delta_theta is computed by
+#          exp(jfk * delta_theta) * pose = targeted_pose
+def compute_delta_theta(jfk, theta, targeted_pose, use_body_jacobian):
+    jac, poses = jfk(theta)
+    pose_inv = SE3_Func.inv(poses[-1])
+    error = (
+        SE3_Func.log(
+            SE3_Func.compose(pose_inv, targeted_pose)
+            if use_body_jacobian
+            else SE3_Func.compose(targeted_pose, pose_inv)
+        )
+        .view(-1, 6, 1)
+        .view(-1, 6, 1)
+    )
+    return (jac[-1].pinverse() @ error).view(-1, robot.dof), error.norm().item()
+
+
+# The following function runs IK. We first create random joint angles to compute
+# a set of target poses for the end effector. Then, starting from zero joint angles,
+# we try to recover joint angles that can achieve the target pose, by iteratively
+# updating the joint angles using the body/spatial jacobian. Note that IK has infinite
+# solutions, so, while we are able to recover the target poses after a few iterations,
+# there is no guarantee that we can recover the joint angles used to generate these
+# poses in the first place.
+def run_ik_using_XYZ_method(batch_size, step_size, use_body_jacobian):
+    target_theta = torch.rand(batch_size, robot.dof, dtype=dtype)
+    target_pose: torch.Tensor = fk(target_theta)[0]
+    theta_opt = torch.zeros_like(target_theta)
+    for _ in range(50):
+        delta_theta, error = compute_delta_theta(
+            jfk_b, theta_opt, target_pose, use_body_jacobian
+        )
+        print(error)
+        if error < 1e-4:
+            break
+        theta_opt = theta_opt + step_size * delta_theta
+
+
+print("---------------------------------------------------")
+print("Body Jacobian")
+print("---------------------------------------------------")
+run_ik_using_XYZ_method(100, 0.5, use_body_jacobian=True)
+
+print("---------------------------------------------------")
+print("Spatial Jacobian")
+print("---------------------------------------------------")
+run_ik_using_XYZ_method(10, 0.2, use_body_jacobian=True)
+
+
+# *********************************************
+# ** INVERSE KINEMATICS AS LS OPTIMIZATION
+# *********************************************
+print("---------------------------------------------------")
+print("Theseus Optimizer")
+print("---------------------------------------------------")
+
+
+# IK can also be solved as an optimization problem:
+#      min \|log(pose^-1 * targeted_pose)\|^2
+# as the following
+def targeted_pose_error(optim_vars, aux_vars):
```

Comment:
Would this be similar to the [Local](https://github.com/facebookresearch/theseus/blob/main/theseus/embodied/misc/local_cost_fn.py#L14)/Difference CostFunction on SE3?